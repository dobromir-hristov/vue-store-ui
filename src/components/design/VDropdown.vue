<template>
  <div
    v-clickaway="close"
    class="VDropdown"
  >
    <slot
      name="trigger"
      :toggleVisibility="toggleVisibility"
    />
    <div
      v-if="isVisible"
      class="VDropdown__dropdown"
    >
      <slot name="content"/>
    </div>
  </div>
</template>

<script>
import { directive } from 'vue-clickaway'

/**
 * Super basic dropdown. Would use popper.js or something more serious...
 * @module VDropdown
 */
export default {
  name: 'VDropdown',
  directives: { Clickaway: directive },
  data () {
    return {
      isVisible: false
    }
  },
  methods: {
    toggleVisibility () {
      this.isVisible = !this.isVisible
    },
    close () {
      this.isVisible = false
    }
  }
}
</script>
