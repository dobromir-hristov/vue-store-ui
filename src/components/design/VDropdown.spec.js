import VDropdown from './VDropdown'

/**
 * @param overrides
 * @returns {{config, wrapper: Wrapper<V> | Wrapper<Vue> | *}}
 */
const createWrapper = (overrides) => {
  let config = _merge(createComponentMocks(),
    {
      scopedSlots: {
        trigger: '<a @click="props.toggleVisibility" data-testid="trigger">CLICK</a>'
      },
      slots: {
        content: 'DROPDOWN_CONTENT'
      }
    },
    overrides
  )
  const wrapper = mount(VDropdown, config)
  return {
    config,
    wrapper
  }
}

describe('VDropdown', () => {
  it('renders the VDropdown with a trigger', () => {
    const { wrapper } = createWrapper()
    expect(wrapper.find(testid('trigger')).exists()).toBeTruthy()
  })

  it('allows toggling the dropdown on/off', () => {
    const { wrapper } = createWrapper()
    wrapper.find(testid('trigger')).trigger('click')
    expect(wrapper.text()).toContain('DROPDOWN_CONTENT')
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.vm.close()
    expect(wrapper.text()).not.toContain('DROPDOWN_CONTENT')
  })
})
