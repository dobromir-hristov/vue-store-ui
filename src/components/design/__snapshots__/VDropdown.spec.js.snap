// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VDropdown allows toggling the dropdown on/off 1`] = `
<div class="VDropdown"><a data-testid="trigger">CLICK</a>
  <div class="VDropdown__dropdown">DROPDOWN_CONTENT</div>
</div>
`;
